### PR TITLE
feat(levm): ef tests from eip5656: MCOPY

### LINKED ISSUE #749: LEVM: (EF Tests) EIP5656 MCOPY
To remove from ignored this suites:
- `"stEIP5656-MCOPY".into(),`
- `"eip5656_mcopy".into(),`

### PR DESCRIPTION
**Pull request by iinaki**
**Motivation**
This [EIP](https://eips.ethereum.org/EIPS/eip-5656) introduces the `MCOPY` instruction.

The tests failed because we were letting them to overflow in some arithmetic operations, this caused the memory expansion to not be handled correctly, as it could panic.

**Changes**

- To remove from ignored this suites:
    - "stEIP5656-MCOPY".into(),
    - "eip5656_mcopy".into(),
- Added checks to those arithmetic ops that overflowed.

**Description**

Closes #749

### CODE REVIEW

File: crates/vm/levm/tests/ef_tests.rs [hunk 1]
```diff
@@ -46,8 +46,8 @@ fn get_ignored_groups() -> HashSet<String> {
         "stEIP158Specific".into(),
         "stEIP4844-blobtransactions".into(),
         "eip4844_blobs".into(),
-        "stEIP5656-MCOPY".into(),
-        "eip5656_mcopy".into(),
+        // "stEIP5656-MCOPY".into(),
+        // "eip5656_mcopy".into(),
```
Comment (ilitteri) [inline crates/vm/levm/tests/ef_tests.rs:50]: Why are we commenting these?

**Review by ilitteri — COMMENTED** [review]
Looks good, left a couple of questions.

File: crates/vm/levm/src/memory.rs
[the remark this answers is not in the document]
Comment (iinaki) [reply crates/vm/levm/src/memory.rs:1]: Basically last_memory_size_word and new_memory_size_word are used to calculate the memory expansion cost by takin the difference between the previous and the current size of the words, they can't be made constants as they depent on the `memory_byte_size` that is variable.
I did take out 512 as a constant, thanks!

File: crates/vm/levm/tests/ef_tests.rs (hunk 1 above)
Comment (iinaki) [reply crates/vm/levm/tests/ef_tests.rs]: Removed! Didn't noticed I left those!

File: crates/vm/levm/src/memory.rs [hunk 2]
```diff
@@ -64,16 +65,25 @@ impl Memory {
...
-            (new_memory_size_word * new_memory_size_word) / 512 + (3 * new_memory_size_word);
+
+        let new_memory_size_word = memory_byte_size
+            .checked_add(WORD_SIZE - 1)
+            .ok_or(VMError::OverflowInArithmeticOp)?
+            / WORD_SIZE;
+
+        let new_memory_cost = new_memory_size_word
+            .checked_mul(new_memory_size_word)
+            .and_then(|square| square.checked_div(MEMORY_EXPANSION_QUOTIENT))
```
Comment (juanimedone) [inline crates/vm/levm/src/memory.rs:80]: This `checked_div` is not necessary as `MEMORY_EXPANSION_QUOTIENT` is not 0.

Comment (iinaki) [reply crates/vm/levm/src/memory.rs]: Done!

File: crates/vm/levm/src/memory.rs [hunk 3]
```diff
@@ -64,16 +65,25 @@ impl Memory {
...
+        let new_memory_cost = new_memory_size_word
+            .checked_mul(new_memory_size_word)
+            .map(|square| square / MEMORY_EXPANSION_QUOTIENT)
+            .and_then(|cost| cost.checked_add(new_memory_size_word.checked_mul(3)?))
+            .ok_or(VMError::OverflowInArithmeticOp)?;
         let last_memory_size_word = (self.data.len() + WORD_SIZE - 1) / WORD_SIZE;
-        let last_memory_cost =
-            (last_memory_size_word * last_memory_size_word) / 512 + (3 * last_memory_size_word);
-        new_memory_cost - last_memory_cost
+        let last_memory_cost = (last_memory_size_word * last_memory_size_word)
```
Comment (juanimedone) [inline crates/vm/levm/src/memory.rs:84]: You could add a `checked_mul` here too.

File: crates/vm/levm/src/memory.rs [hunk 4]
```diff
@@ -64,16 +65,25 @@ impl Memory {
...
+            / MEMORY_EXPANSION_QUOTIENT
+            + (3 * last_memory_size_word);
```
Comment (juanimedone) [inline crates/vm/levm/src/memory.rs:86]: And a `checked_add` here.

File: crates/vm/levm/src/memory.rs (hunk 3 above)
Comment (iinaki) [reply crates/vm/levm/src/memory.rs]: Done!

File: crates/vm/levm/src/memory.rs (hunk 4 above)
Comment (iinaki) [reply crates/vm/levm/src/memory.rs]: Done!

File: crates/vm/levm/src/opcode_handlers/stack_memory_storage_flow.rs [hunk 5]
```diff
@@ -242,9 +246,21 @@ impl VM {
...
 
-        let words_copied = (size + WORD_SIZE - 1) / WORD_SIZE;
-        let memory_byte_size = (src_offset + size).max(dest_offset + size);
-        let memory_expansion_cost = current_call_frame.memory.expansion_cost(memory_byte_size);
+        // let words_copied = (size + WORD_SIZE - 1) / WORD_SIZE;
+        let words_copied = size
+            .checked_add(WORD_SIZE - 1)
+            .ok_or(VMError::OverflowInArithmeticOp)?
+            / WORD_SIZE;
+        // let memory_byte_size = (src_offset + size).max(dest_offset + size);
```
Comment (juanimedone) [inline crates/vm/levm/src/opcode_handlers/stack_memory_storage_flow.rs:254]:
```suggestion
        let words_copied = size
            .checked_add(WORD_SIZE - 1)
            .ok_or(VMError::OverflowInArithmeticOp)?
            / WORD_SIZE;
```

**Review by juanimedone — APPROVED** [review]

File: crates/vm/levm/src/opcode_handlers/stack_memory_storage_flow.rs (hunk 5 above)
Comment (iinaki) [reply crates/vm/levm/src/opcode_handlers/stack_memory_storage_flow.rs]: Done!

**Pull request merged**